File: app/src/test/java/com/lifetrace/execute/data/repository/TaskWireMapperTest.kt
package com.lifetrace.execute.data.repository

import com.lifetrace.execute.domain.task.ExecutionTask
import com.lifetrace.execute.domain.task.ExecutionTaskPriority
import com.lifetrace.execute.domain.task.ExecutionTaskStatus
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class TaskWireMapperTest {
    @Test
    fun roundTripPreservesTaskFieldsAndUsesAcceptedServerVersion() {
        val task = ExecutionTask(
            id = "task-1",
            userId = "user-1",
            title = "Write tests",
            description = "Cover the production mapper",
            projectId = "project-1",
            status = ExecutionTaskStatus.IN_PROGRESS,
            priority = ExecutionTaskPriority.HIGH,
            dueAt = "2026-09-02T12:00:00Z",
            scheduledAt = "2026-09-01T10:00:00Z",
            completedAt = null,
            createdAt = "2026-09-01T08:00:00Z",
            updatedAt = "2026-09-01T09:00:00Z",
            localVersion = 4,
            serverVersion = "11",
            modifiedByDevice = "device-a",
        )

        val payload = TaskWireMapper.toPayload(task)
        val decoded = TaskWireMapper.fromPayload(payload, serverVersion = "12")

        assertEquals(task.copy(serverVersion = "12"), decoded)
        val root = JSONObject(payload)
        assertEquals("task-1", root.getJSONObject("meta").getString("id"))
        assertEquals("project-1", root.getString("projectId"))
    }

    @Test
    fun nullableFieldsSurviveRoundTrip() {
        val task = ExecutionTask(
            id = "task-2",
            userId = "user-1",
            title = "Offline task",
            createdAt = "2026-09-01T08:00:00Z",
            updatedAt = "2026-09-01T08:00:00Z",
            localVersion = 1,
        )

        val decoded = TaskWireMapper.fromPayload(
            TaskWireMapper.toPayload(task),
            serverVersion = "1",
        )

        assertNull(decoded.description)
        assertNull(decoded.projectId)
        assertNull(decoded.dueAt)
        assertNull(decoded.scheduledAt)
        assertNull(decoded.completedAt)
        assertEquals(ExecutionTaskStatus.TODO, decoded.status)
        assertEquals(ExecutionTaskPriority.NORMAL, decoded.priority)
    }
}

File: flutter_app/test/task_wire_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lifetrace_execute/data/repository/task_wire_mapper.dart';
import 'package:lifetrace_execute/domain/task/execution_task.dart';

void main() {
  test('task wire mapper preserves Compose Sync v1 payload shape', () {
    const task = ExecutionTask(
      id: 'task-1',
      userId: 'user-1',
      title: 'Write tests',
      description: 'sync parity',
      projectId: 'project-1',
      status: ExecutionTaskStatus.inProgress,
      priority: ExecutionTaskPriority.high,
      dueAt: '2026-09-11T12:00:00.000Z',
      scheduledAt: '2026-09-11T10:00:00.000Z',
      createdAt: '2026-09-10T10:00:00.000Z',
      updatedAt: '2026-09-10T11:00:00.000Z',
      localVersion: 4,
      serverVersion: '12',
      modifiedByDevice: 'device-1',
    );

    final payload = TaskWireMapper.toPayload(task);
    final meta = payload['meta'] as Map<String, dynamic>;

    expect(meta['id'], 'task-1');
    expect(meta['userId'], 'user-1');
    expect(meta['deletedAt'], isNull);
    expect(meta['localVersion'], 4);
    expect(meta['serverVersion'], '12');
    expect(payload['status'], 'in_progress');
    expect(payload['priority'], 'high');
    expect(payload['projectId'], 'project-1');

    final restored = TaskWireMapper.fromPayload(payload, serverVersion: '13');
    expect(restored.id, task.id);
    expect(restored.title, task.title);
    expect(restored.status, task.status);
    expect(restored.priority, task.priority);
    expect(restored.localVersion, task.localVersion);
    expect(restored.serverVersion, '13');
  });
}
